include/geode/basic/database.h

```diff
@@ -0,0 +1,164 @@
+
+        /*!
+         * Classe holding a const reference of a data.
+         * @warning Do not destroy this Data class before the const reference
+         * obtained using its get() method is no longer in used
+         */
+        class opengeode_basic_api Data
+        {
+        public:
+            Data( std::shared_ptr< Storage > storage );
+            ~Data();
+            Data( Data&& other );
+
+            template < typename DataType >
+            const DataType& get()
+            {
+                const auto* typed_data =
+                    dynamic_cast< const DataType* >( &data() );
+                OPENGEODE_EXCEPTION(
+                    typed_data, "[Data::get] Cannot cast data into DataType" );
+                return *typed_data;
+            }
+
+        private:
+            const Identifier& data() const;
+
+        private:
+            IMPLEMENTATION_MEMBER( impl_ );
+        };
+
+    public:
+        Database( absl::string_view directory );
+        ~Database();
+
+        index_t nb_data() const;
+
+        template < typename DataType >
+        uuid register_new_data( DataType&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            uuid new_id;
+            register_unique_data(
+                new_id, absl::make_unique< DataType >( std::move( data ) ) );
+            return new_id;
+        }
+
+        template < typename DataType >
+        uuid register_new_data( std::unique_ptr< DataType >&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            uuid new_id;
+            register_unique_data( new_id, std::move( data ) );
+            return new_id;
+        }
+
+        template < typename DataType >
+        void update_data( const uuid& id, DataType&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            return register_unique_data(
+                id, absl::make_unique< DataType >( std::move( data ) ) );
+        }
+
+        template < typename DataType >
+        void update_data( const uuid& id, std::unique_ptr< DataType >&& data )
+        {
+            static_assert( std::is_base_of< Identifier, DataType >::value,
+                "[Database::register_data] Data is not a subclass of "
+                "Identifier" );
+            return register_unique_data( id, std::move( data ) );
+        }
+
+        void delete_data( const uuid& id );
+
+        /*!
+         * Retrieve a read only reference to the data corresponding to the given
+         * uuid.
+         */
+        Data get_data( const uuid& id ) const;
```

Comment:
read only but not const ref?

Reply:
Yes, this is a wrapper, not the real data